### DISCUSSION TITLE
MacOS App should detect running emacs brew service and connect as a GUI emacsclient.

### DISCUSSION THREAD

**Discussion by cwlbraa** [d12frosted/homebrew-emacs-plus#716] [General]
I love that this formula includes a brew service, and for the longest time I was running that service alongside the Emacs.App. I only recently realized that these 2 paths to invoking Emacs are unaware of each other, and the App will not attempt to connect to the daemon started by the brew service. 

If this worked, it'd be amazing for startup times and keeping a single persistent Emacs daemon.

**Comment by d12frosted** [comment]
Hey @cwlbraa 

I think this kind of relates to https://github.com/d12frosted/homebrew-emacs-plus/issues/158. That issue stagnated though.

What do you think about the following solution? 

In addition to `Emacs.app`, `emacs-plus` provides extra application `Emacs Client.app` that is a wrapper around `emacsclient`. This way we don't need to come up with fancy heuristics, but instead rely on `emacsclient` itself that knows how to work. 

The only possible issue I see - there will be no way to configure socket and server, you'll have to use defaults.

**Comment by cwlbraa** [comment]
I'd use the simple solution if it existed, so yeah I think it's good. I'd probably hide the full Emacs.app from Alfred and end up with basically the same behavior as the fancy thing I previously described. I'd also be perfectly happy with defaults.

**Comment by shaunsingh** [comment]
Currently I use an "application" created via automator that runs `emacsclient -c`, its been working great for me so far 

The solution you propose seems about the same, having it included with emacs-plus would be very nice

**Comment by C1oud555** [comment]
It would be great to have the `Emacs Client.app`.

**Comment by diocletiann** [comment]
I'm having trouble finding `Emacs Client.app`, do you have to install with a specific flag? Thanks.

**Comment by d12frosted** [comment]
@restfuladi you can't find it because it doesn't exist :) But there is `emacsclient` binary installed that you can use for your needs.

**Comment by diocletiann** [comment]
> @restfuladi you can't find it because it doesn't exist :) But there is `emacsclient` binary installed that you can use for your needs.

Makes sense, thanks. Any ideas on how to integrate that with Spotlight/Alfred?

**Comment by gf3** [comment]
@restfuladi Spotlight/Alfred have issues with symlinks, however [Raycast](https://www.raycast.com) works really well!

**Comment by mjgiffin** [comment]
> > @restfuladi you can't find it because it doesn't exist :) But there is `emacsclient` binary installed that you can use for your needs.
> 
> Makes sense, thanks. Any ideas on how to integrate that with Spotlight/Alfred?

If you create an Automator application for emacs client and put it in /Applications, that should work with Spotlight. I recommend full path:
/opt/homebrew/bin/emacsclient -c -a emacs "$@" >/dev/null 2>&1

**Comment by rross101** [comment]
Piggy-backing on this issue, happy to make another one - when I launch emacsclient from terminal or Alfred, it doesn't switch to the app - the window opens behind other windows, and then I have to manually switch to it. Has anyone a fix for that?

**Comment by ChauhanT** [comment]
Hi, I think Emacsclient.app is a fantastic idea. Is this something that will become available in the homebrew installation? I am new to macos (coming from linux), and I have very little experience doing mac-y things. I also observed what @rross101 said. It's a minor inconvenience, but of course, if it can be addressed it's even better. 

Software architecture of the mac is so, so inefficient. It's like homebrew is creating an OS within an OS, and obviously there are no standards like XDG or how permissions are managed etc. I can't even get my brew-installed hunspell to integrate with emacs-plus. Sorry for the rant, but macos is just such a user unfriendly environment to work in if you are trying to something even remotely technical.

**Comment by d12frosted** [comment]
> Hi, I think Emacsclient.app is a fantastic idea. Is this something that will become available in the homebrew installation?

The idea is to provide this as part of Emacs+ installation. But it's not high on my priority list. So if someone really wants it, contributions are welcome.

> Software architecture of the mac is so, so inefficient. It's like homebrew is creating an OS within an OS, and obviously there are no standards like XDG or how permissions are managed etc.

Indeed, macOS is slightly (quite?) different in many regards and might be confusing if you come from newer Linux-es. macOS is based on older FreeBSD.

> I can't even get my brew-installed hunspell to integrate with emacs-plus. Sorry for the rant, but macos is just such a user unfriendly environment to work in if you are trying to something even remotely technical.

Look, if you need some help, don't hesitate to open a [discussion][1]. The problem you describe is not really related to Emacs+, but it's somewhere on the border, so maybe it's fine :)

[1]: https://github.com/d12frosted/homebrew-emacs-plus/discussions

**Comment by citizen428** [comment]
@ChauhanT This is what I did:

1. Start the emacs-plus service and register it to launch on login/boot:
    ```sh
    brew services start emacs-plus@29
   ```
2. Create an "application" with Automator: 
    <img width="1074" alt="Screenshot 2022-12-28 at 16 14 28" src="https://user-images.githubusercontent.com/47985/209788484-6ddbbb73-e6e4-4f3b-bc28-6e3ec1c88098.png">
3. If you want, change the app icon (Right click -> "Get info" or `Cmd-i`). I used the same I compile emacs-plus with, `nobu417-big-sur-icon`:
     <img width="109" alt="Screenshot 2022-12-28 at 16 17 49" src="https://user-images.githubusercontent.com/47985/209788828-003bdf2c-dec4-474c-999c-891bdcc83cf5.png">

@rross101 You could use Apple Script for that: `osascript -e 'tell application "Emacs" to activate'` You can wrap that up in aliases or shell functions, e.g. 

```sh
function ec() {
  emacsclient $@ && osascript -e 'tell application "Emacs" to activate'
}
```

**Reply by andreacfromtheapp** [reply]
this works amazingly for me :) thanks!

**Comment by johnhamelink** [comment]
I found I had to tweak @citizen428's answer a little bit to get the applescript to work reliably:

<img width="627" alt="Screenshot 2023-04-10 at 20 33 57" src="https://user-images.githubusercontent.com/101739/230982255-5dc7f281-5219-4f5a-8076-6f56e1e284cd.png">

**Comment by rross101** [comment]
> I found I had to tweak @citizen428's answer a little bit to get the applescript to work reliably:
> 
> <img alt="Screenshot 2023-04-10 at 20 33 57" width="627" src="https://user-images.githubusercontent.com/101739/230982255-5dc7f281-5219-4f5a-8076-6f56e1e284cd.png">

This works for me. Thank you!

**Comment by GauravTalreja** [comment]
I changed it to following to set Emacs Client as the default text editor:

<img width="431" alt="image" src="https://user-images.githubusercontent.com/59036640/236397212-7d3829b0-67ec-45f2-a30d-64213b51de9e.png">

**Comment by michaelssingh** [comment]
I just checked quickly and noticed that these Automator scripts do not launch emacs when the server is not running.  Has anyone else encountered this?

**Reply by rafaelcn** [reply]
I'm a new macOS user and on linux there MUST be an emacs process running as a daemon to use `emacsclient` at all so I guess it's expected to not launch emacs if there are no emacs running as a daemon.

**Comment by rafaelcn** [comment]
hello, I come from the linux world and just recently I got a macOS (running the last sequoia version). I've had a ton of problems trying to run `emacsclient`, lastly I tried the automator solution with the application running as a service (`brew service start ....`).

The service is up and running and whenever I try the automator recipe it kind of opens an emacs frame but it doesn't. On the dock there's the emacs icon but whenever I click it nothing happens, I tried right clicking it and clicking on the _show all windows option_ but then it shows this.

![image](https://github.com/user-attachments/assets/5c6a68b0-101e-4342-b158-40a31bb41ed8)

what can you guys make of this? running as a daemon I tried a unix socket and now I set my emacs to start a TCP socket.

```lisp
;; to anyone wondering what I meant by using a TCP socket.
(when (eq system-type 'darwin)
    (sets server-use-tcp t))
```

**Reply by d12frosted** [reply]
It should be working out of the box. You do `brew services start emacs-plus@30` and then `emacsclient` connects without any issues. Just make sure you don't have any extra socket arguments passed (unless you known what you are doing). Also check that you are not setting `EMACS_SERVER_FILE` env variable - i.e. do not use TCP :) As per TCP usage - I don't know if it even works on macos.

**Reply by rafaelcn** [reply]
Yep, can confirm. I just set the `server-use-tcp` to `nil` and it worked after reloading the emacs service. The only thing that I noticed is a bit off is the color of the window border, which is like a gray color instead of the usual white one.

**Reply by rafaelcn** [reply]
Nevermind, that's isn't true.

**Comment by d12frosted** [comment]
Closing as we already have Emacs Client.app bundled both in formula and cask.
